src/smc-webapp/paying_and_upgrading.cjsx

```diff
@@ -0,0 +1,470 @@
+        ###
+        state =
+            upgrading : true
+
+        limits = @get_quota_limits()
+        console.log(JSON.stringify(limits))
+        for name, data of @props.quota_params
+            factor = data.display_factor
+            if name == 'network' or name == 'member_host'
+                limit = if limits.limits[name] > 0 then 1 else 0
+                if maximize
+                    current_value = limit ? limits.current[name]
+                else
+                    current_value = 0
+            else
+                if maximize
+                    current_value = limits.limits[name] ? limits.current[name]
+                else
+                    current_value = 0
+            state["upgrade_#{name}"] = misc.round2(current_value * factor)
+            upgrades = {}
+        return state
+
+    max_upgrades : ->
+        @setState(@getUpgraderState(true))
+
+    reset_upgrades : ->
+        @setState(@getUpgraderState(false))
+
+    show_upgrade_quotas : ->
+        @setState(upgrading : true)
+
+    cancel_upgrading : ->
+        state =
+            upgrading : false
+
+        current = @props.upgrades_you_applied_to_this_project
+
+        for name, data of @props.quota_params
+            factor = data.display_factor
+            current_value = current[name] ? 0
+            state["upgrade_#{name}"] = misc.round2(current_value * factor)
+
+        @setState(state)
+
+    is_upgrade_input_valid : (input, max) ->
+        val = misc.parse_number_input(input, round_number=false)
+        if not val? or val > Math.max(0, max)
+            return false
+        else
+            return true
+
+    # the max button will set the upgrade input box to the number given as max
+    render_max_button : (name, max) ->
+        <Button
+            bsSize  = 'xsmall'
+            onClick = {=>@setState("upgrade_#{name}" : max)}
+            style   = {padding:'0px 5px'}
+        >
+            Max
+        </Button>
+
+    render_reset_button : (name) ->
+        <Button
+            bsSize  = 'xsmall'
+            onClick = {=>@setState("upgrade_#{name}" : 0)}
+            style   = {padding:'0px 5px'}
+        >
+            Reset
+        </Button>
+
+    render_addon : (misc, name, display_unit, limit, val, remaining) ->
+        <div style={minWidth:'81px'}>{"#{misc.plural(2,display_unit)}"} {@render_max_button(name, limit) if remaining > 0}{@render_reset_button(name) if val > 0}</div>
+
+    render_get_more_of_this_upgrade : (name) ->
+        # Detrimine if user has subscription that has given upgrade. Like cores requires permium subscription
+        <DropdownButton title='Get more of this upgrade by' bsSize='xsmall'>
+            <MenuItem>Purchasing an additional subscription</MenuItem>
+            <MenuItem>Downgrading other projects</MenuItem>
+            <MenuItem>Resetting this upgrade type for all projects</MenuItem>
+        </DropdownButton>
```

Comment:
I still don't know what this DropdownButton really does. it makes this look a bit chaotic, since it repeats so often. maybe for now, we should leave that out?